scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+from collections.abc import Iterable
+import numpy as np
+from scipy._lib._util import _asarray_validated
+from scipy.linalg import block_diag, LinAlgError
+from .lapack import _compute_lwork, get_lapack_funcs
+
+__all__ = ['cossin']
+
+
+def cossin(X, p=None, q=None, minus_upper=True, compute_u=True,
+           compute_vh=True):
+    """
+    Compute the cosine-sine (CS) decomposition of an orthogonal/unitary matrix.
+
+    X is an ``(m, m)`` orthogonal/unitary matrix, partitioned as the following
+    where upper left block has the shape of ``(p, q)``::
+
+                                   ┌                   ┐
+                                   │ I  0  0 │ 0  0  0 │
+        ┌           ┐   ┌         ┐│ 0  C  0 │ 0 -S  0 │┌         ┐*
+        │ X11 │ X12 │   │ U1 │    ││ 0  0  0 │ 0  0 -I ││ V1 │    │
+        │ ────┼──── │ = │────┼────││─────────┼─────────││────┼────│
+        │ X21 │ X22 │   │    │ U2 ││ 0  0  0 │ I  0  0 ││    │ V2 │
+        └           ┘   └         ┘│ 0  S  0 │ 0  C  0 │└         ┘
+                                   │ 0  0  I │ 0  0  0 │
+                                   └                   ┘
+
+    ``U1``, ``U2``, ``V1``, ``V2`` are square orthogonal/unitary matrices and
+    ``C`` and ``S`` are ``(r, r)`` nonnegative diagonal matrices satisfying
+    ``C^2 + S^2 = I`` where ``r = min(p, m-p, q, m-q)``.
+
+    Moreover, the rank of the identity matrices are ``min(p, q) - r``,
+    ``min(p, m - q) - r``, ``min(m - p, q) - r``, and ``min(m - p, m - q) - r``
+    respectively.
+
+    X can be supplied either by itself or by its subblocks in an iterable. See
+    the examples below.
+
+    Parameters
+    ----------
+    X : array_like, iterable
+        complex unitary or real orthogonal matrix to be decomposed, or iterable
+        of subblocks ``X11``, ``X12``, ``X21``, ``X22``, when ``p``, ``q`` are
+        omitted.
+    p : int
+        Number of rows of the upper left block ``X11``
+    q : int
+        Number of columns of the upper left block ``X11``
+    minus_upper : bool, optional
```

Comment:
I thought this is done via lapack itself via the signs keyword. Also it seems to me that swap_sign or something like that would be a better candidate for the name.

Reply:
Yeah, the signs keyword takes care of the `U` and `VH` matrices but `theta` stays the same - how we have to take this into consideration ourselves when we construct the `CS` matrix.